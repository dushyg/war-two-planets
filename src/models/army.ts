export interface Army {
  name: string;
  forces: Map<string, number>;
}
