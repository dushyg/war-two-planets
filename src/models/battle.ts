export interface Battle {
  remaining: number;
  uncountered: number;
  used: number;
}
