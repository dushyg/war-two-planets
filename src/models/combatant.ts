export interface Combatant {
  /**
   * Count of combatants in this category
   */
  numberOfUnits: number;
  /**
   * Number of units of opposition that this 1 unit can defeat
   */
  tacklingPower: number;
  /**
   * Deployment Position from left to right amoungst other type of combatants within its own army
   */
  deploymentPosition: number;

  /**
   * Code Name of combatant like 'H', 'E', 'AT', or 'SG'
   */
  codeName: string;
}
