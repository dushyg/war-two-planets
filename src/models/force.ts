export interface Force {
  name: string;
  /**
   * The code name with which it is passed as an input to the app
   */
  code: string;
  /**
   * Number of enemy units that can be countered with 1 of this unit
   */
  power: number;
}
