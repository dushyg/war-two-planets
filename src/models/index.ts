export { Army } from './army';
export { Battle } from './battle';
export { Force } from './force';
export { Task } from './task';
