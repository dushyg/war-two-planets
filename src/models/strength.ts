export interface Strength {
  /**
   * Count of combatants in this category
   */

  numberOfUnits: number;
  /**
   * Number of units of opposition that this 1 unit can defeat
   */
  tacklingPower: number;
  /**
   * Deployment Position from left to right amoungst other type of combatants within its own army
   */
  deploymentPosition: number;
}
