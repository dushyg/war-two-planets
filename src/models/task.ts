export enum Task {
  'Power',
  'RankForRank',
  'Substitution',
}
