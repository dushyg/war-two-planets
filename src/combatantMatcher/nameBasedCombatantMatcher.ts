import { Combatant, Army } from '../models';
import { CombatantMatcher } from './combatantMatcher';
import { FORCE_CODES } from '../constants';

/**
 * Implements the Like for Like rule and finds defender based on codeName
 */
export class NameBasedCombatantMatcher implements CombatantMatcher {
  public getDefender(invader: Combatant, defenderArmy: Army): Combatant {
    const defender = defenderArmy.forces.get(invader.codeName);
    if (defender) {
      return defender;
    }
    return {
      codeName: FORCE_CODES.NONE,
      deploymentPosition: -1,
      numberOfUnits: 0,
      tacklingPower: 0,
    } as Combatant;
  }
}
