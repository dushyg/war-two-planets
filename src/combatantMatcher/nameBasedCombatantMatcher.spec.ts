import { NameBasedCombatantMatcher } from './nameBasedCombatantMatcher';
import { Combatant, Army } from '../models';

describe('NameBasedCombatantMatcher', () => {
  it('getDefender should return default object when no match is found', () => {
    const combatant: Combatant = new NameBasedCombatantMatcher().getDefender(
      {} as Combatant,
      { forces: new Map<string, Combatant>() } as Army,
    );

    expect(combatant).toMatchSnapshot();
  });
});
