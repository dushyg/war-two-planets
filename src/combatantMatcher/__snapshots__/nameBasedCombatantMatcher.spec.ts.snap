// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameBasedCombatantMatcher getDefender should return default object when no match is found 1`] = `
Object {
  "codeName": "NONE",
  "deploymentPosition": -1,
  "numberOfUnits": 0,
  "tacklingPower": 0,
}
`;
